### ISSUE TITLE
8-bit and 4-bit indexed texture viewing tabs

### ISSUE THREAD

**Issue by ManDude** [cbhacks/CrashEdit#95]
Kind of like PVV does:
- Left-clicking anywhere with the "Palette Select" tool will select a row of 16 pixels in 4-bit mode, and 256 in 8-bit.
- Left-clicking anywhere with the "Texture Select" tool will select a box. Size of the box can be changed by right-clicking (in short, left-clicking anchors the box and right-click or holding the right mouse button lets you change its size).
- A panel below the 15-bit texture chunk image (this panel is present in the 8-bit and 4-bit tabs, yes, and is where you do the palette and texture selection) shows the output image after the correct colors have been applied.
- "Save as [whatever]" option that saves the output image as whatever.
- Zoom button(s).

**Comment by warenhuis** [comment]
Are you guys sure that the BGR555 option gives the right colors for the 8- and 4-bit textures? I reversed-engineered a CLUT-line for the '?' Box in chunk Cr10T,  and can not find the right 'CLUT' line in the chunk.

**Comment by ughman** [comment]
> Are you guys sure that the BGR555 option gives the right colors for the 8- and 4-bit textures?

I'm not really sure of that, no. The only reason I "know" the colors are "right" is because they do appear correct for the 16-bit universal logo from crash 2's title screen 3C.NSF.

**Comment by warenhuis** [comment]
Ok, it is right, I found the line for the '?' Box with the help of PVV and it matched the lines in the BGR555 option, but it was in an unexpected order. Now I will try to rethink how the CLUT-id's are ordered, because my own line was in a chronological order (of the data, 0 to 15) which did not work.
![clut](https://cloud.githubusercontent.com/assets/15270225/12699115/482463b2-c7b0-11e5-8355-f7f7bde23171.png)

**Comment by ManDude** [comment]
They're not ordered in any specific way, so you'll just have to try out everything.

**Comment by ughman** [comment]
According to this page http://problemkaputt.de/psx-spx.htm#gpurenderingattributes the CLUT attribute for the GPU textured poly commands has the low 6 bits as the X coordinate, and the next 9 bits as the Y coordinate (plus one bit unused). The key is the X coordinate is in 16-texel (32-byte) units, so cluts are always 16-texel aligned.

Not sure if that's what you're talking about, though. Just trying to help.

**Comment by warenhuis** [comment]
I was trying to find out how the 16-bit pixels inside a CLUT line were ordered. Sorry if I was ambigious on that (I found out now, so no problem). I do have some comments on textures however, these attribute settings sound familiar with what I found in the textureinfo of the models in CB2.

But I figured that one doesn't actually need the GPU if all we want is the colors, as all required references are inside the same texture chunk (the one exception is when a 8-bit texture is specified in section '4' (each chunk has 4 sections, 4 is the rightmost one) with relative UV-offsets that are bigger than 128, in which case the selectionregion shifts to the next chunk in VRAM), but for now I'm going to assume that such a specification never occurs in the texinfo.

The GPU will become important when we need all the different colormodes.

Thanks for the help though.
